File: benchmark/zig_benchmark/src/test_internal.zig
const std = @import("std");
const hash_zig = @import("hash-zig");

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    std.debug.print("Testing Zig internal signing and verification...\n", .{});

    // Initialize the scheme
    var scheme = try hash_zig.GeneralizedXMSSSignatureScheme.init(allocator, .lifetime_2_8);
    defer scheme.deinit();

    // Generate keypair
    std.debug.print("Generating keypair...\n", .{});
    var keypair = try scheme.keyGen(0, 256);
    defer keypair.secret_key.deinit();

    // Test message
    const test_message = "Hello, Zig internal test!";
    const epoch: u32 = 0;

    // Convert message to bytes
    var message_bytes: [32]u8 = undefined;
    @memset(&message_bytes, 0);
    @memcpy(message_bytes[0..@min(test_message.len, 32)], test_message);

    // Sign the message
    std.debug.print("Signing message...\n", .{});
    const signature = try scheme.sign(keypair.secret_key, epoch, message_bytes);
    defer signature.deinit();

    // Verify the signature
    std.debug.print("Verifying signature...\n", .{});
    const is_valid = try scheme.verify(&keypair.public_key, epoch, message_bytes, signature);

    std.debug.print("Result: {}\n", .{is_valid});
    
    if (is_valid) {
        std.debug.print("✅ SUCCESS: Zig internal signing and verification works!\n", .{});
    } else {
        std.debug.print("❌ FAILED: Zig internal verification failed!\n", .{});
    }
}

File: benchmark/rust_benchmark/src/test_internal.rs
use hashsig::signature::generalized_xmss::instantiations_poseidon_top_level::lifetime_2_to_the_8::SIGTopLevelTargetSumLifetime8Dim64Base8;
use hashsig::signature::{SignatureScheme, SignatureSchemeSecretKey};
fn main() {
    println!("Testing Rust internal signing and verification...");

    let mut rng = rand::rng();
    let (pk, mut sk) = SIGTopLevelTargetSumLifetime8Dim64Base8::key_gen(&mut rng, 0, 256);

    // Test message
    let test_message = "Hello, Rust internal test!";
    let epoch = 0u32;

    // Convert message to bytes (truncate/pad to 32 bytes)
    let mut message_bytes = [0u8; 32];
    let message_bytes_slice = test_message.as_bytes();
    let copy_len = std::cmp::min(message_bytes_slice.len(), 32);
    message_bytes[..copy_len].copy_from_slice(&message_bytes_slice[..copy_len]);

    println!("Generating keypair...");
    println!("Signing message...");

    // Prepare the secret key for the epoch
    while !sk.get_prepared_interval().contains(&(epoch as u64)) {
        sk.advance_preparation();
    }

    // Sign the message
    let signature_result =
        SIGTopLevelTargetSumLifetime8Dim64Base8::sign(&sk, epoch, &message_bytes);

    let signature = match signature_result {
        Ok(sig) => sig,
        Err(e) => {
            println!("❌ FAILED: Rust signing failed: {:?}", e);
            return;
        }
    };

    println!("Verifying signature...");

    // Verify the signature
    let is_valid =
        SIGTopLevelTargetSumLifetime8Dim64Base8::verify(&pk, epoch, &message_bytes, &signature);

    println!("Result: {}", is_valid);

    if is_valid {
        println!("✅ SUCCESS: Rust internal signing and verification works!");
    } else {
        println!("❌ FAILED: Rust internal verification failed!");
    }
}
